archspec/cpu/cpuid.py

```diff
@@ -0,0 +1,697 @@
+                )
+
+                # EBX
+                ebx = self._run_asm(
+                    instruction,  # mov ax,0x8000000?
+                    b"\x0f\xa2"  # cpuid
+                    b"\x89\xD8"  # mov ax,bx
+                    b"\xC3",  # ret
+                )
+
+                # ECX
+                ecx = self._run_asm(
+                    instruction,  # mov ax,0x8000000?
+                    b"\x0f\xa2"  # cpuid
+                    b"\x89\xC8"  # mov ax,cx
+                    b"\xC3",  # ret
+                )
+
+                # EDX
+                edx = self._run_asm(
+                    instruction,  # mov ax,0x8000000?
+                    b"\x0f\xa2"  # cpuid
+                    b"\x89\xD0"  # mov ax,dx
+                    b"\xC3",  # ret
+                )
+
+                # Combine each of the 4 bytes in each register into the string
+                for reg in [eax, ebx, ecx, edx]:
+                    for n in [0, 8, 16, 24]:
+                        processor_brand += chr((reg >> n) & 0xFF)
+
+        # Strip off any trailing NULL terminators and white space
+        processor_brand = processor_brand.strip("\0").strip()
+
+        return processor_brand
+
+
+def _get_cpu_info_from_cpuid_actual():
+    """
+    Warning! This function has the potential to crash the Python runtime.
+    Do not call it directly. Use the get_cpu_info_from_cpuid function instead.
+    It will safely call this function in another process.
+    """
+
+    info = {}
+
+    try:
+        # Return none if this is not an X86 CPU
+        if not _is_x86():
+            return {}
+
+        cpuid = CPUID()
+
+        # Get the cpu info from the CPUID register
+        max_extension_support = cpuid.get_max_extension_support()
+        info = cpuid.get_info()
+
+        processor_brand = cpuid.get_processor_brand(max_extension_support)
+
+        info = {
+            "vendor_id_raw": cpuid.get_vendor_id(),
+            "brand_raw": processor_brand,
+            "stepping": info["stepping"],
+            "model": info["model"],
+            "family": info["family"],
+            "processor_type": info["processor_type"],
+            "flags": cpuid.get_flags(max_extension_support),
+        }
+
+    except Exception as err:
+        return {}
+
+    return info
+
+
+def _get_cpu_info_from_cpuid_subprocess_wrapper(queue):
+    output = _get_cpu_info_from_cpuid_actual()
+    queue.put(_obj_to_b64(output))
+
+
+def get_cpu_info_from_cpuid():
+    """
+    Returns the CPU info gathered by querying the X86 cpuid register in a new process.
+    Returns {} on non X86 cpus.
+    """
+
+    from multiprocessing import Process, Queue
+
+    # Return {} if this is not an X86 CPU
+    if not _is_x86():
+        return {}
+
+    try:
+        # Start running the function in a subprocess
+        queue = Queue()
+        p = Process(target=_get_cpu_info_from_cpuid_subprocess_wrapper, args=(queue,))
+        p.start()
```

Comment:
Why do we need to spawn a new process at all?  Can't we encode a much simpler function, detached from the `py-cpuinfo` implementation?